jsrc/algorithm.hpp

```diff
@@ -1,39 +1,119 @@
-
 #include <algorithm>
+#include <array>
 #include <tuple>
 
 namespace algo {
 
+/**
+ * @brief Variadic version of `std::equal_to`
+ */
+static auto constexpr is_equal_to = [](auto const &first, auto const &...values) -> bool {
+    static_assert(sizeof...(values) != 0);
+    return ((first == values) && ...);
+};
+
+/**
+ * @brief Variadic version of `std::not_equal_to`
+ */
+static auto constexpr is_not_equal_to = [](auto const &first, auto const &...values) -> bool {
+    return !is_equal_to(first, values...);
+};
+
 /**
  * @brief Variadic version of `std::mismatch`
  */
 template <typename P, typename I, typename... Is>
-[[nodiscard]] auto
-zip_find(P pred, I f, I l, Is... fs) {
+[[nodiscard]] static auto constexpr zip_find(P const &pred, I f, I l, Is... fs) {
+    static_assert(sizeof...(Is) != 0);
     while (f != l) {
         if (pred(*f, *fs...)) break;
         ++f, ((void)++fs, ...);
     }
     return std::tuple{f, fs...};
 }
 
+/**
+ * @brief Variadic version of `std::mismatch`, for ranges.
+ * @note This can be the same name as zip_find if we use concepts.
+ */
+template <typename P, typename R, typename... Rs>
+[[nodiscard]] static auto constexpr zip_find_r(P const &pred, R const &r, Rs const &...rs) {
```

Comment:
Let's just put this in a nested namespace and drop the `_r`. `algo::ranges::` probably to be consistent with C++ Ranges.